helix-core/src/main/java/org/apache/helix/manager/zk/CallbackEventTPFactory.java

```diff
@@ -0,0 +1,96 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+// returns a thread pool object given a HelixManager identifier
+public class CallbackEventTPFactory {
+  static private final ReadWriteLock _lock = new ReentrantReadWriteLock();
+  static private Map<Integer, ThreadPoolExecutor> _managerToCallBackThreadPoolMap = new HashMap();
+  static private Map<Integer, AtomicInteger> _callBackEventProcessorCountPerThreadPool =
+      new HashMap();
+
+  public static ThreadPoolExecutor getOrCreateThreadPool(int managerHash) {
+    // should not use general lock for read
+    _lock.readLock().lock();
+    ThreadPoolExecutor result = null;
+    if (_managerToCallBackThreadPoolMap.containsKey(managerHash)) {
+      result = _managerToCallBackThreadPoolMap.get(managerHash);
+      _callBackEventProcessorCountPerThreadPool.get(managerHash).incrementAndGet();
+    }
+    _lock.readLock().unlock();
+    if (result == null) {
+      result = getOrCreateThreadPoolHelper(managerHash);
+    }
+    return result;
+  }
+
+  private static ThreadPoolExecutor getOrCreateThreadPoolHelper(int managerHash) {
+    ThreadPoolExecutor result;
+    _lock.writeLock().lock();
+    // first check if the key is already in the map
+    if (_managerToCallBackThreadPoolMap.containsKey(managerHash)) {
+      // downgrade to read lock since we dont need to modify the map
+      _lock.readLock().lock();
+      result = _managerToCallBackThreadPoolMap.get(managerHash);
+      _lock.readLock().unlock();
+    } else {
+      ThreadFactory namedThreadFactory = new ThreadFactoryBuilder()
+          .setNameFormat(String.format("CallbackHandlerExecutorService - %s ", managerHash))
+          .build();
+      result = new ThreadPoolExecutor(5, 5, 1, TimeUnit.SECONDS, new LinkedBlockingQueue<>(),
```

Comment:
On my second thought, I think we should not keep so many threads. Otherwise, if the manager has only one callback for a path, we will still create so many threads. This does not make sense.
One reasonable way is to set the core thread count same as the required callback count. This can be derived from the HelixMananger listener fields. However, there seems to be no easy way to get the number directly even we have the manager object.

Let's think more about it.

Reply:
The reason why I am using ThreadPoolExecutor(5, 5, 1, TimeUnit.SECONDS,...) is exactly the same. Since we set `allowCoreThreadTimeOut ` to true, thread will be terminated if not used after 1 sec.

Reply:
" when the number of threads is greater than the core, this is the maximum time that excess idle threads will wait for new tasks before terminating."

Since the core thread is also 5, it will keep 5 live threads in the pool.

Reply:
Seems to be a very strange design, why not just set the core thread to be 0? I'm just curious. The logic seems to be working.

Reply:
From java Doc:
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html#allowsCoreThreadTimeOut()
```If there are more than corePoolSize but less than maximumPoolSize threads running, a new thread will be created only if the queue is full.
```
If have an unbounded queue when core = 1, and max = 5, all new tasks (except the first one) will be pushed to the queue and thread count is always 1. I think the only way to have a dynamically increasing/decreasing thread pool with unbounded queue is having a coreThread==maxThread==X and timeToLive = Y.
However, the thread pool will have 0 idle threads after the timeout has expired. So I decided to change back to a fixSize thread pool.

Reply:
I cannot believe that I never noticed this before. Thanks for raising this up. I think the original solution will work. But some details need another thought to refine.

Alternatively, we can also do as this post said, https://stackoverflow.com/questions/15485840/threadpoolexecutor-with-unbounded-queue-not-creating-new-threads
I have tested the solution and it works great. Admittedly, it is more complex. So up to you : )

Reply:
From our offline discussion. I will keep the same but update the TTL and thread count.